File: androidApp/app/src/main/java/com/harnest/app/engine/HttpBridge.kt
package com.harnest.app.engine

import android.util.Log
import okhttp3.Call
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * HTTP bridge — mirrors HttpBridge.ets: executes the fetch requested by the
 * QuickJS side (via onFetch), streaming headers/chunks back as fetch events.
 * Pending device_camera photos are attached to the next chat/completions
 * request as image_url parts (multimodal vision).
 */
class HttpBridge(private val emit: (fetchId: Int, kind: String, a: String, b: String) -> Unit) {

    companion object {
        private const val TAG = "HttpBridge"
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(300, TimeUnit.SECONDS)
        .build()

    private val active = mutableMapOf<Int, Call>()
    private val lock = Any()

    /** requestJson: {url, method, headers, body} */
    fun start(fetchId: Int, requestJson: String) {
        val req = try {
            JSONObject(requestJson)
        } catch (e: Throwable) {
            emit(fetchId, "fail", "bad fetch request json: ${e.message}", "")
            return
        }
        val url = req.optString("url", "")
        val method = req.optString("method", "GET").uppercase()
        val headers = req.optJSONObject("headers")
        val bodyStr = when {
            !req.has("body") || req.isNull("body") -> null
            else -> req.optString("body")
        }

        val vision = attachPendingImages(url, bodyStr)
        val effectiveUrl = vision?.first ?: url
        val effectiveBody = vision?.second ?: bodyStr
        val attachedImages = vision != null

        val call = client.newCall(buildRequest(effectiveUrl, method, headers, effectiveBody))
        synchronized(lock) { active[fetchId] = call }
        call.enqueue(object : okhttp3.Callback {
            override fun onFailure(call: Call, e: IOException) {
                synchronized(lock) { active.remove(fetchId) }
                Log.w(TAG, "fetch#$fetchId fail: ${e.message}")
                emit(fetchId, "fail", e.message ?: "network error", "")
            }

            override fun onResponse(call: Call, response: Response) {
                var resp = response
                if (attachedImages && resp.code in 400..499 && bodyStr != null) {
                    try {
                        Log.w(TAG, "fetch#$fetchId: vision request rejected (${resp.code}), retrying without images")
                        resp.close()
                        resp = client.newCall(buildRequest(url, method, headers, bodyStr)).execute()
                    } catch (e: Throwable) {
                        synchronized(lock) { active.remove(fetchId) }
                        emit(fetchId, "fail", "vision retry failed: ${e.message}", "")
                        return
                    }
                }
                try {
                    val headerJson = JSONObject()
                    for ((k, v) in resp.headers) {
                        if (!headerJson.has(k)) headerJson.put(k, v)
                    }
                    emit(fetchId, "headers", resp.code.toString(), headerJson.toString())
                    val utf8 = Utf8StreamDecoder()
                    resp.body?.byteStream()?.use { stream ->
                        val buffer = ByteArray(16 * 1024)
                        while (true) {
                            val n = stream.read(buffer)
                            if (n < 0) break
                            if (n > 0) {
                                val text = utf8.decode(buffer, n)
                                if (text.isNotEmpty()) emit(fetchId, "chunk", text, "")
                            }
                        }
                    }
                    val rest = utf8.flush()
                    if (rest.isNotEmpty()) emit(fetchId, "chunk", rest, "")
                    emit(fetchId, "done", "", "")
                } catch (e: Throwable) {
                    emit(fetchId, "fail", e.message ?: "read error", "")
                } finally {
                    synchronized(lock) { active.remove(fetchId) }
                    resp.close()
                }
            }
        })
    }

    private fun buildRequest(url: String, method: String, headers: JSONObject?, bodyStr: String?): Request {
        val builder = Request.Builder().url(url)
        headers?.keys()?.forEach { k -> builder.header(k, headers.optString(k, "")) }
        val mediaType = "application/json; charset=utf-8".toMediaType()
        when (method) {
            "GET" -> builder.get()
            "POST" -> builder.post((bodyStr ?: "").toRequestBody(mediaType))
            "PUT" -> builder.put((bodyStr ?: "").toRequestBody(mediaType))
            "DELETE" -> if (bodyStr != null) builder.delete(bodyStr.toRequestBody(mediaType)) else builder.delete()
            "PATCH" -> builder.patch((bodyStr ?: "").toRequestBody(mediaType))
            "HEAD" -> builder.head()
            else -> builder.method(method, bodyStr?.toRequestBody(mediaType))
        }
        return builder.build()
    }

    /**
     * Append pending camera photos to the next chat/completions request as a
     * user message with OpenAI-compatible image_url content parts. When the
     * configured provider is text-only (e.g. zhipu coding endpoint), reroute
     * the request to a vision-capable route. Returns (url, body) when images
     * were attached, or null to leave the request untouched.
     */
    private fun attachPendingImages(url: String, bodyStr: String?): Pair<String, String>? {
        if (bodyStr.isNullOrEmpty()) return null
        if (!url.contains("/chat/completions")) return null
        if (!VisionAttach.hasPending()) return null
        val body = try {
            JSONObject(bodyStr)
        } catch (e: Throwable) {
            return null
        }
        val messages = body.optJSONArray("messages") ?: return null
        val images = VisionAttach.drain()
        if (images.isEmpty()) return null
        for (i in 0 until messages.length()) {
            val m = messages.optJSONObject(i) ?: continue
            if (m.has("content") && m.isNull("content")) m.put("content", "")
        }
        val content = JSONArray().put(
            JSONObject().put("type", "text").put(
                "text",
                "[系统附图] 以下为用户刚通过 device_camera 拍摄的照片（按拍摄顺序）。请直接查看图片内容进行分析或回答，不要凭空猜测画面："
            )
        )
        for (img in images) {
            content.put(
                JSONObject().put("type", "image_url")
                    .put("image_url", JSONObject().put("url", img))
            )
        }
        messages.put(JSONObject().put("role", "user").put("content", content))
        body.put("messages", messages)
        var targetUrl = url
        if (url.contains("open.bigmodel.cn")) {
            if (url.contains("/api/coding/")) {
                targetUrl = "https://open.bigmodel.cn/api/paas/v4/chat/completions"
                body.put("model", "glm-4.6v")
                Log.i(TAG, "vision route: coding endpoint is text-only, rerouting to glm-4.6v on standard endpoint")
            } else if (!body.optString("model").endsWith("v")) {
                body.put("model", "glm-4.6v")
                Log.i(TAG, "vision route: switching model to glm-4.6v")
            }
        }
        Log.i(TAG, "attaching ${images.size} photo(s) to chat request as image_url parts")
        return targetUrl to body.toString()
    }

    fun abortAll() {
        synchronized(lock) {
            active.values.forEach { it.cancel() }
            active.clear()
        }
    }
}

File: iosApp/Sources/Service/HttpBridge.swift
import Foundation

/// Streaming HTTP for harness.js — mirrors HttpBridge.kt / http_bridge.cpp.
/// - {url, method, headers, body} → headers/chunk/done/fail events
/// - SSE-friendly chunked text via URLSession.AsyncBytes + UTF-8 boundary-safe decoding
/// - pending camera photos attach as image_url parts on /chat/completions (with
///   vision rerouting for open.bigmodel.cn coding endpoints)
final class HttpBridge {

    private let emit: (Int, String, String, String) -> Void
    private let session: URLSession
    private let lock = NSLock()
    private var active: [UUID: Task<Void, Never>] = [:]

    init(emit: @escaping (Int, String, String, String) -> Void) {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 300
        cfg.timeoutIntervalForResource = 3600
        session = URLSession(configuration: cfg)
        self.emit = emit
    }

    func start(fetchId: Int, requestJson: String) {
        let uuid = UUID()
        let task = Task { [weak self] in
            await self?.run(uuid: uuid, fetchId: fetchId, requestJson: requestJson)
            _ = self?.lock.withLock { self?.active.removeValue(forKey: uuid) }
        }
        lock.lock()
        active[uuid] = task
        lock.unlock()
    }

    func abortAll() {
        lock.lock()
        let tasks = Array(active.values)
        active.removeAll()
        lock.unlock()
        tasks.forEach { $0.cancel() }
    }

    private func run(uuid: UUID, fetchId: Int, requestJson: String) async {
        guard let req = (try? JSONSerialization.jsonObject(with: Data(requestJson.utf8))) as? [String: Any],
              let urlStr = req["url"] as? String,
              let base = URL(string: urlStr) else {
            emit(fetchId, "fail", "bad fetch request", "")
            return
        }
        let method = ((req["method"] as? String) ?? "GET").uppercased()
        let headers = (req["headers"] as? [String: Any])?.compactMapValues { $0 as? String } ?? [:]
        let body = req["body"] as? String

        let vision = Self.attachPendingImages(url: urlStr, body: body)
        let effectiveUrl = vision?.0 ?? urlStr
        let effectiveBody = vision?.1 ?? body
        let attachedImages = vision != nil

        let request = Self.buildRequest(
            urlStr: effectiveUrl, fallback: base, method: method,
            headers: headers, body: effectiveBody
        )

        do {
            var (bytes, response) = try await session.bytes(for: request)
            var status = (response as? HTTPURLResponse)?.statusCode ?? 0

            // vision fallback: 4xx with attached images → retry original request
            if attachedImages && (400...499).contains(status) && body != nil {
                let retry = Self.buildRequest(
                    urlStr: urlStr, fallback: base, method: method,
                    headers: headers, body: body
                )
                let (b2, r2) = try await session.bytes(for: retry)
                bytes = b2
                response = r2
                status = (r2 as? HTTPURLResponse)?.statusCode ?? 0
            }

            if let http = response as? HTTPURLResponse {
                var h: [String: String] = [:]
                for (k, v) in http.allHeaderFields {
                    if let ks = k as? String {
                        h[ks] = String(describing: v)
                    }
                }
                emit(fetchId, "headers", String(status), Self.encode(h))
            } else {
                emit(fetchId, "headers", "0", "{}")
            }

            var decoder = Utf8StreamDecoder()
            var iterator = bytes.makeAsyncIterator()
            var buf: [UInt8] = []
            buf.reserveCapacity(16 * 1024)
            while let byte = try await iterator.next() {
                buf.append(byte)
                if buf.count >= 16 * 1024 {
                    if let text = decoder.decode(Data(buf)) {
                        emit(fetchId, "chunk", text, "")
                    }
                    buf.removeAll(keepingCapacity: true)
                }
            }
            if !buf.isEmpty, let text = decoder.decode(Data(buf)) {
                emit(fetchId, "chunk", text, "")
            }
            if let tail = decoder.flush() {
                emit(fetchId, "chunk", tail, "")
            }
            emit(fetchId, "done", "", "")
        } catch {
            if Task.isCancelled { return }
            emit(fetchId, "fail", error.localizedDescription, "")
        }
    }

    private static func buildRequest(
        urlStr: String, fallback: URL, method: String,
        headers: [String: String], body: String?
    ) -> URLRequest {
        var request = URLRequest(url: URL(string: urlStr) ?? fallback)
        request.httpMethod = method
        var contentType: String? = nil
        for (k, v) in headers {
            let lk = k.lowercased()
            if lk == "content-type" { contentType = v; continue }
            if lk == "content-length" || lk == "host" { continue }
            request.setValue(v, forHTTPHeaderField: k)
        }
        if let b = body {
            request.httpBody = Data(b.utf8)
            if contentType == nil { contentType = "application/json; charset=utf-8" }
        }
        if let ct = contentType {
            request.setValue(ct, forHTTPHeaderField: "Content-Type")
        }
        request.timeoutInterval = 300
        return request
    }

    /// Pending camera photos → multimodal user message appended to /chat/completions body.
    private static func attachPendingImages(url: String, body: String?) -> (String, String)? {
        guard let bodyStr = body, !bodyStr.isEmpty else { return nil }
        guard url.contains("/chat/completions") else { return nil }
        guard VisionAttach.hasPending() else { return nil }
        guard var payload = (try? JSONSerialization.jsonObject(with: Data(bodyStr.utf8))) as? [String: Any],
              var messages = payload["messages"] as? [[String: Any]] else { return nil }
        let images = VisionAttach.drain()
        guard !images.isEmpty else { return nil }
        for i in messages.indices {
            if messages[i]["content"] is NSNull {
                messages[i]["content"] = ""
            }
        }
        var content: [[String: Any]] = [[
            "type": "text",
            "text": "[系统附图] 以下为用户刚通过 device_camera 拍摄的照片（按拍摄顺序）。请直接查看图片内容进行分析或回答，不要凭空猜测画面：",
        ]]
        for img in images {
            content.append(["type": "image_url", "image_url": ["url": img]])
        }
        messages.append(["role": "user", "content": content])
        payload["messages"] = messages
        var targetUrl = url
        if url.contains("open.bigmodel.cn") {
            if url.contains("/api/coding/") {
                targetUrl = "https://open.bigmodel.cn/api/paas/v4/chat/completions"
                payload["model"] = "glm-4.6v"
            } else if !((payload["model"] as? String) ?? "").hasSuffix("v") {
                payload["model"] = "glm-4.6v"
            }
        }
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let out = String(data: data, encoding: .utf8) else { return nil }
        return (targetUrl, out)
    }

    private static func encode(_ dict: [String: String]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: dict) else { return "{}" }
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}

/// Chunk-boundary-safe UTF-8 decoder: incomplete multibyte sequences carry to the next chunk.
struct Utf8StreamDecoder {
    private var carry: [UInt8] = []

    mutating func decode(_ data: Data) -> String? {
        var bytes = carry + [UInt8](data)
        carry.removeAll()
        let cut = Self.validPrefixLength(bytes)
        if cut < bytes.count {
            carry = Array(bytes[cut...])
            bytes = Array(bytes[0..<cut])
        }
        if bytes.isEmpty { return nil }
        return String(data: Data(bytes), encoding: .utf8)
    }

    /// End of stream: lossy-decode any leftover bytes (invalid → U+FFFD).
    mutating func flush() -> String? {
        let bytes = carry
        carry.removeAll()
        guard !bytes.isEmpty else { return nil }
        if let s = String(data: Data(bytes), encoding: .utf8) { return s }
        return String(decoding: bytes, as: UTF8.self)
    }

    /// Longest valid-prefix length: scan back ≤4 bytes for a start byte; if the
    /// trailing sequence is incomplete, cut before its start byte.
    private static func validPrefixLength(_ b: [UInt8]) -> Int {
        let n = b.count
        var i = n - 1
        var back = 0
        while i >= 0 && back <= 3 {
            let c = b[i]
            if c & 0xC0 == 0x80 {
                if back == 3 { return max(n - 3, 0) }  // >3 trailing continuations: malformed
                back += 1
                i -= 1
                continue
            }
            let need: Int
            if c >= 0xF0 { need = 4 }
            else if c >= 0xE0 { need = 3 }
            else if c >= 0xC0 { need = 2 }
            else { need = 1 }
            return (n - i) >= need ? n : i
        }
        return max(i + 1, 0)
    }
}
